### PR TITLE
Add shape_idx to CollisionObject2D mouse_entered signal

### PR DESCRIPTION
**Pull request by sygi**
See [proposal](https://github.com/godotengine/godot-proposals/issues/2487) for discussion

### CODE REVIEW

File: scene/main/viewport.h [hunk 1]
```diff
@@ -271,7 +272,10 @@ class Viewport : public Node {
 	bool handle_input_locally = true;
 	bool local_input_handled = false;
 
+	// Collider to frame
 	Map<ObjectID, uint64_t> physics_2d_mouseover;
+	// Collider & shape to frame
+	Map<Pair<ObjectID, int>, uint64_t, PairSort<ObjectID, int>> physics_2d_shape_mouseover;
```
Comment (pouleyKetchoupp) [inline scene/main/viewport.h:278]: Why is `PairSort<ObjectID, int>` needed? Does it fail to compile otherwise?

Comment (sygi) [reply scene/main/viewport.h:278]: Yes. You need to define a comparator operator for Map to know how to hash the objects.

Comment (pouleyKetchoupp) [reply scene/main/viewport.h:278]: Ah yes, that makes sense.

File: doc/classes/CollisionObject2D.xml [hunk 2]
```diff
@@ -234,14 +234,28 @@
...
-				Emitted when the mouse pointer enters any of this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
+				Emitted when the mouse pointer enters any of this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set. Note that moving between different shapes within a single [code]CollisionObject2D[/code] won't cause this signal to be emitted.
 			</description>
 		</signal>
 		<signal name="mouse_exited">
 			<description>
-				Emitted when the mouse pointer exits all this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
+				Emitted when the mouse pointer exits all this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set. Note that moving between different shapes within a single [code]CollisionObject2D[/code] won't cause this signal to be emitted.
+			</description>
+    </signal>
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:244]:
```suggestion
		</signal>
```

File: doc/classes/CollisionObject2D.xml [hunk 3]
```diff
@@ -234,14 +234,28 @@
...
+		<signal name="mouse_shape_entered">
+			<argument index="0" name="shape_idx" type="int">
+			</argument>
+			<description>
+        Emitted when the mouse pointer enters any of this object's shapes or moves from one shape to another. [code]shape_idx[/code] is the child index of the newly entered Shape2d. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:249]:
```suggestion
				Emitted when the mouse pointer enters any of this object's shapes or moves from one shape to another. [code]shape_idx[/code] is the child index of the newly entered [Shape2D]. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
```

File: doc/classes/CollisionObject2D.xml [hunk 4]
```diff
@@ -234,14 +234,28 @@
...
 			</description>
 		</signal>
+		<signal name="mouse_shape_exited">
+			<argument index="0" name="shape_idx" type="int">
+			</argument>
+			<description>
+				Emitted when the mouse pointer exits any of this object's shapes. [code]shape_idx[/code] is the child index of the exited Shape2d. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
+			</description>
+    </signal>
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:258]:
```suggestion
		</signal>
```

File: scene/main/viewport.cpp [hunk 5]
```diff
@@ -2607,6 +2633,7 @@ void Viewport::_drop_mouse_focus() {
 void Viewport::_drop_physics_mouseover(bool p_paused_only) {
 	physics_has_last_mousepos = false;
 
+	// TODO: deduplicate this with the erasing in ss2d.
```
Comment (Calinou) [inline scene/main/viewport.cpp:2636]: What does SS2D mean? I'd replace it with the full expanded text.

**Review by Calinou — CHANGES_REQUESTED** [review]

File: scene/main/viewport.cpp (hunk 5 above)
Comment (sygi) [reply scene/main/viewport.cpp]: ss2d is an object in this file. I added better pointers (but still want to point where in the method the place I point to is, as the method is quite long).

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
It looks great! It seems to works fine, and I don't have much to comment about the code except for the suggestion related to your TODO message.

File: scene/main/viewport.cpp [hunk 6]
```diff
@@ -2607,6 +2633,7 @@ void Viewport::_drop_mouse_focus() {
 void Viewport::_drop_physics_mouseover(bool p_paused_only) {
 	physics_has_last_mousepos = false;
 
+	// TODO: deduplicate this with the erasing in _process_picking (send to 2D).
```
[the remark this answers is not in the document]
Comment (sygi) [reply scene/main/viewport.cpp]: Thanks for the comment. Unfortunately, it's not as simple as you suggest. I refactored the code while keeping the logic as is; do you know if I can further simplify it? ie. do you know why the shapes are removed only when they are castable to CollisionObject2D in one case and in all cases in the other?

On a more general note, is there a place for documentation of private methods where I should document this better than with a comment in .h?

[the remark this answers is not in the document]
Comment (pouleyKetchoupp) [reply scene/main/viewport.cpp]: After looking at the code, it looks like it should erase from the list outside of the CollisionObject2D check in both cases to me.

The difference was introduced in #45775.
@RandomShaper Do you remember if there was a specific reason for this difference or if it could be a typo?

For more context, the original logic when triggering exit events is:
https://github.com/godotengine/godot/blob/277db14b647d3184b341a359020c0acc4549afaf/scene/main/viewport.cpp#L735-L742

But in `_drop_physics_mouseover`, `to_erase.push_back(E)` is done only after casting to `CollisionObject2D`:
https://github.com/godotengine/godot/blob/277db14b647d3184b341a359020c0acc4549afaf/scene/main/viewport.cpp#L2604-L2614

If we can confirm they both should be like the first case, it will be possible to further simplify the code, and it would need a fix on 3.x branch as well.

@sygi About documenting private methods, there's no specific place for that. You can leave extra comments in the cpp, either right before or within the method if you want to add more details.

[the remark this answers is not in the document]
Comment (RandomShaper) [reply scene/main/viewport.cpp]: The second case is different because because of `p_paused_only`. If it's `true`, the node shall not be removed.

That said, I think it may be good to erase it as well if the check for a valid instance fails (the `if (o)` check).

Furthermore, refactoring the code so if `p_paused_only` is checked in an outer scope may help, by allowing to have a loop similar to the one in the first snippet (unconditional removal). However, an additional variant (that checks `co->can_process()`) will still be needed, so I'm not sure if that could help with what is being done in this PR.

[the remark this answers is not in the document]
Comment (sygi) [reply scene/main/viewport.cpp]: Could you clarify if the logic in #528d029 is correct or not? (it's keeps `p_paused_only`)

[the remark this answers is not in the document]
Comment (RandomShaper) [reply scene/main/viewport.cpp]: I think it isn't.

`_drop_physics_mouseover()` is called with `p_paused_only = true` every frame, to send `_mouse_exit` to nodes that have just gotten paused. So, if `to_erase.push_back(E);` happens regardless `p_paused_only && co->can_process()`, any node that gets hovered will immediately be removed from being considered hovered, so it won't get `_mouse_exit` when unhovering actually happens.

In other words, a node should be "unmarked" from being hovered (`to_erase.push_back(E);`) if, and only if, it's as well notified to have been unhovered (`co->_mouse_exit()`).

[the remark this answers is not in the document]
Comment (sygi) [reply scene/main/viewport.cpp]: Thanks for the comment. Sorry about that, but I'm still a bit confused. You say:
> it may be good to erase it as well if the check for a valid instance fails (the if (o) check)

In the code above I only moved the `to_erase` check to happen if `!o || !co`. The handling of the case when `p_paused_only && co->can_process()` is the same (no erasure). In particular, under the assumption that `co != null`, `co->_move_exit()` happens iff `to_erase.push_back(E)`.

Are we speaking about splitting handling of `p_paused_only` and `co->can_process`? should we erase when `p_paused_only` even if `!co`?

[the remark this answers is not in the document]
Comment (pouleyKetchoupp) [reply scene/main/viewport.cpp]: Yeah, I think it's fine then, `continue` will prevent removal from the list when not processed yet. Just waiting for confirmation from @RandomShaper, but the current state of the PR should be good to go.

[the remark this answers is not in the document]
Comment (RandomShaper) [reply scene/main/viewport.cpp]: Today, with a pair of fresh eyes, I've been able to read the code correctly. It's fine indeed.

File: scene/main/viewport.cpp [hunk 7]
```diff
@@ -2607,39 +2595,73 @@ void Viewport::_drop_mouse_focus() {
...
+	for (Map<Pair<ObjectID, int>, uint64_t, PairSort<ObjectID, int>>::Element *E = physics_2d_shape_mouseover.front(); E; E = E->next()) {
+		if (!clean_all_frames && E->get() == frame_reference) {
+			continue;
+		}
+
+		Object *o = ObjectDB::get_instance(E->key().first);
+		if (o) {
+			CollisionObject2D *co = Object::cast_to<CollisionObject2D>(o);
+			if (co) {
+				if (p_paused_only && co->can_process()) {
```
Comment (pouleyKetchoupp) [inline scene/main/viewport.cpp:2652]: Should this check be `if (clean_all_frames && p_paused_only && co->can_process())` in order to follow the same logic as for the mouse_exit events?

Comment (sygi) [reply scene/main/viewport.cpp]: Yeah, thanks, let's make them consistent, done.

Note that on the logic level these two have the same behavior as `p_paused_only => clean_all_frames` (I only pass p_paused_only when clean_all_frames is true).

File: scene/main/viewport.h [hunk 8]
```diff
@@ -271,7 +272,12 @@ class Viewport : public Node {
...
+	// Cleans up colliders corresponding to old frames or all of them.
+	void _cleanup_mouseover_colliders(bool clean_all_frames, bool p_paused_only, uint64_t frame_reference = 0);
```
Comment (pouleyKetchoupp) [inline scene/main/viewport.h:280]: clean_all_frames and frame_reference are missing `p_` prefix, sorry I missed it before.

Comment (sygi) [reply scene/main/viewport.h:280]: Done, thanks for catching that. Am I right that the convention is to always append `p_` to method arguments (as opposed to fields/local variables)?

**Review by pouleyKetchoupp — APPROVED** [review]

**Comment by pouleyKetchoupp** [comment]
Great job, thank you!

File: doc/classes/CollisionObject2D.xml [hunk 9]
```diff
@@ -234,12 +234,26 @@
 		</signal>
 		<signal name="mouse_entered">
 			<description>
-				Emitted when the mouse pointer enters any of this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
+				Emitted when the mouse pointer enters any of this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set. Note that moving between different shapes within a single [code]CollisionObject2D[/code] won't cause this signal to be emitted.
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:237]:
```suggestion
				Emitted when the mouse pointer enters any of this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [member collision_layer] bit to be set. Note that moving between different shapes within a single [CollisionObject2D] won't cause this signal to be emitted.
```

File: doc/classes/CollisionObject2D.xml [hunk 10]
```diff
@@ -234,12 +234,26 @@
...
 			</description>
 		</signal>
 		<signal name="mouse_exited">
 			<description>
-				Emitted when the mouse pointer exits all this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
+				Emitted when the mouse pointer exits all this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set. Note that moving between different shapes within a single [code]CollisionObject2D[/code] won't cause this signal to be emitted.
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:242]:
```suggestion
				Emitted when the mouse pointer exits all this object's shapes. Requires [member input_pickable] to be [code]true[/code] and at least one [member collision_layer] bit to be set. Note that moving between different shapes within a single [CollisionObject2D] won't cause this signal to be emitted.
```

File: doc/classes/CollisionObject2D.xml [hunk 11]
```diff
@@ -234,12 +234,26 @@
...
+			</description>
+		</signal>
+		<signal name="mouse_shape_entered">
+			<argument index="0" name="shape_idx" type="int">
+			</argument>
+			<description>
+				Emitted when the mouse pointer enters any of this object's shapes or moves from one shape to another. [code]shape_idx[/code] is the child index of the newly entered [Shape2D]. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:249]:
```suggestion
				Emitted when the mouse pointer enters any of this object's shapes or moves from one shape to another. [code]shape_idx[/code] is the child index of the newly entered [Shape2D]. Requires [member input_pickable] to be [code]true[/code] and at least one [member collision_layer] bit to be set.
```

File: doc/classes/CollisionObject2D.xml [hunk 12]
```diff
@@ -234,12 +234,26 @@
...
+			</description>
+		</signal>
+		<signal name="mouse_shape_exited">
+			<argument index="0" name="shape_idx" type="int">
+			</argument>
+			<description>
+				Emitted when the mouse pointer exits any of this object's shapes. [code]shape_idx[/code] is the child index of the exited Shape2d. Requires [member input_pickable] to be [code]true[/code] and at least one [code]collision_layer[/code] bit to be set.
```
Comment (Calinou) [inline doc/classes/CollisionObject2D.xml:256]:
```suggestion
				Emitted when the mouse pointer exits any of this object's shapes. [code]shape_idx[/code] is the child index of the exited [Shape2D]. Requires [member input_pickable] to be [code]true[/code] and at least one [member collision_layer] bit to be set.
```

**Review by Calinou — CHANGES_REQUESTED** [review]

**Review by Calinou — APPROVED** [review]
Thanks!

**Comment by Calinou** [comment]
Feel free to open a `3.x` pull request if it can be done in a backwards-compatible way :slightly_smiling_face:

**Comment by sygi** [comment]
I don't think I'll be up for it -- looking forward for a transition to a stable 4.0 :)
